torch/_tensor_docs.py

```diff
@@ -1527,6 +1539,22 @@ def add_docstr_all(method, docstr):
   :meth:`Tensor.coalesce` for details.
 """)
 
+add_docstr_all('_indices',
```

Comment:
I am not sure we want to officially document these internal attributes.
I think it's fine to expose them but maybe we just want to create a new public facing API for them?

Something like: `t.indices(coalesced=True)`. Such that by default the Tensor will get coalesced. 
Advanced users that read the doc and know what uncoalesced means can actually ask for `t.indices(coalesced=False)` to get the same thing as `t._indices`?

Reply:
@albanD , the suggestion of replacing `t._indices()` with `t.indices(coalesced=False)` sounds good. Just a small nit: when `t` is already coalesced, the specification `coalesced=False` sounds like the result should be the indices of uncoalesed tensor, which, of course, are not available anymore. So, I suggest: `t.indices(coalesce=False)`. What do you think?

This will eliminate `_indices`, `_values`, also `_nnz` is unnecessary as it is equivalent to`_values(coalesced=False).shape[0]`, in addition, `nnz` ought to be renamed to `nse`.

I suggest a separate PR for the above changes.

Reply:
`coalesce` sounds good yes. I would set the default to `True` though to make sure that there is no surprise for users that don't read the doc in too much details.
A separate PR sounds good. I just don't want to document here things that we don't want exposed.

Reply:
Yes, I agree that `coalesce=True` ought to be the default.

So, the plan is as follows:
- remove `_values`, `_indices`, `_nnz` documentations and any references to them in the docs.
- in a separate PR implement `indices(coalesce=True)` and `values(coalesce=True)` support that includes:
  - document the `coalesce=True` argument
  - replace all occurrences of `_indices/_values` with `indices(coalesce=False)/values(coalesce=False)`
  - when users call `_indices/_values/_nnz`, trigger a deprecation warning
  - implement/document a new method `nse(coalesce=True)` as a replacement of `_nnz`, deprecate the usage of `_nnz`.

Reply:
Sounds good to me.
cc @ngimel who is also interested in sparse Tensors.

Reply:
@albanD @ngimel , I just realized that introducing `indices(coalesce=True)` may be problematic in the sense that `coalesce` is not an inplace operation and it is not clear to me if we should introduce `coalesce_` or not?

For example, consider a pseudocode for `indices`:
```
def indices(self, coalesce=True):
    if coalesce:
        return self.coalesce()._indices()
    return self._indices()
```
This implementation is problematic because each call to `.indices()` will trigger a coalesce process which will be a waste when doing this for the same sparse tensor multiple times.

If we would have inplace `coalesce_`, then the implementation would be:
```
def indices(self, coalesce=True):
    if coalesce:
        self.coalesce_()
    return self._indices()
```

What do you think? Are there any reasons for not introducing in-place coalesce?

Reply:
> This implementation is problematic because each call to .indices() will trigger a coalesce process which will be a waste when doing this for the same sparse tensor multiple times.

This if true if you assume that getting the indices should be a super cheap op. But in general, you can consider getting the indices "expensive" and the user should not access the indices for the same sparse Tensor in a for loop!
I am not sure how I feel about `t.indices()` modifying `t` inplace. That does not really sound right.

The idea for `t.indices()` in my mind is for an inexperienced user to be able to get the indices from a sparse Tensor in a safe way without weird side effect or unexpected results. I don't think there is a strict performance constraint on this.
For advanced users that know about coalescing, they already do and will keep doing the right thing of either coalescing before getting the indices or asking for uncoalesced indices if that's what they want.

Reply:
I agree with @albanD, `indices` should not modify `t`. As for whether we should have in-place `coalesce_` method, I'm open-minded, maybe we should. For dense tensors we have `transpose_()`, `permute_()`, `resize_()` etc, so why not. 

Docs should list that `indices` is not a cheap call. We already have cheap `is_coalesced()`, right? So `indices` could check that and skip coalescing if possible. People who want to avoid constant coalescing calls can call coalesce once  before making a series of `indices` calls.

Reply:
Yes, `is_coalesced()` is cheap.

I have created https://github.com/pytorch/pytorch/pull/45695 that introduces methods `indices(coalesce=True)`, `values(coalesce=True)`, and `nse(coalesce=True)`.

I will resolve this conversation and let's continue it in https://github.com/pytorch/pytorch/pull/45695 .

Leaving introduction of `coalesce_` open as it is out of scope of this update-docs-PR.